Add enable-render-process-reuse flag

rzhao271 opened the pull request:
Affects #120431

This PR adds an `enable-render-process-reuse` flag to the runtime arguments in `argv.json` (one can access that file by running the `workbench.action.configureRuntimeArguments` command. By default, the runtime argument is `false`, but one can set it to `true` in argv.json.

deepak1556 reviewed (changes requested):
Can you reuse the runtime arguments code path, check `enable-browser-code-loading` for example and add a new flag `enable-render-process-reuse`.

rzhao271 commented:
The flag is now `enable-render-process-reuse`, which takes in either `true` or `false`. Its default value is `true`.
It's set in `argv.json`. I was having trouble testing it earlier, until I realized this. 😅

deepak1556 commented on src/main.js:
```diff
@@ -223,10 +223,18 @@ function configureCommandlineSwitchesSync(cliArgs) {
 					break;
+
+				case 'enable-render-process-reuse':
+					if (argvValue === false) {
+						enableRenderProcessReuse = false;
+					}
```
```suggestion
					if (argvValue === false) {
						app.allowRendererProcessReuse = false;
					} else {
					        app.allowRendererProcessReuse = true;
					}
```

deepak1556 commented:
Sanity Check before merging,

* Workbench process ID does not change on reload
* Swapping workspaces in a window also keeps the same process
* Ensure child process from workbench get restarted
   * Extension Host
   * Watcher Service
   * Search Service
* Smoke test https://github.com/Microsoft/vscode/wiki/Smoke-Test

- [ ] Windows @rzhao271 
- [ ] Linux @deepak1556 
- [ ] MacOS @bpasero

rzhao271 commented:
Current issues on Windows:

1. The search service isn't reloading
2. The smoke tests don't run

rzhao271 commented:
It turns out the search service not loading until a search is made is expected behaviour.
Rob points out that extension activation could cause a search to occur, which I confirmed on Insiders: I launched code-insiders with all extensions disabled on an untrusted workspace. Then, I opened the process explorer, and confirmed that searchService wasn't there.

rzhao271 commented:
Current progress:

On Mac, sometimes a hang occurs when restarting a workspace. When there are no hangs, the smoke tests pass, too. This is with process reuse enabled on Electron 11. At least there's no more process leaks.

On Windows, Electron 12 needs to be merged in to bring in https://github.com/electron/electron/pull/28175, as described here: https://github.com/microsoft/vscode/issues/120431#issuecomment-816354849. Considering how it should be merged in relatively soon, I'll check whether it's been merged in tomorrow. If it hasn't, I'll just reapply these changes locally on the Electron 12 branch and run the sanity checks from there.

bpasero commented on src/main.js:
```diff
@@ -25,9 +25,7 @@ const { getUserDataPath } = require('./vs/platform/environment/node/userDataPath
-// Disable render process reuse, we still have
-// non-context aware native modules in the renderer.
-app.allowRendererProcessReuse = false;
+app.allowRendererProcessReuse = true;
```
Shouldn't this remain `false` for now and only turn into `true` when the user has opted in via `enable-render-process-reuse` below?

rzhao271 replied:
Changed feature to opt-in.

bpasero commented on src/main.js:
```diff
@@ -168,7 +166,10 @@ function configureCommandlineSwitchesSync(cliArgs) {
-		'log-level'
+		'log-level',
+
+		// Enables render process reuse. Default value is 'true'. See https://github.com/electron/electron/issues/18397
```
I would have thought that the default remains `false` and this flag is to enable reuse selectively.

rzhao271 replied:
Changed feature to opt-in.

bpasero commented on src/vs/workbench/services/search/electron-browser/searchService.ts:
```diff
@@ -84,6 +86,10 @@ export class DiskSearch implements ISearchResultProvider {
 		const channel = getNextTickChannel(client.getChannel('search'));
 		this.raw = new SearchChannelClient(channel);
+
+		this.lifecycleService.onDidShutdown(() => {
```
Don't we need similar code for the file watcher too?

rzhao271 replied:
I did another sanity check:
Assuming that `src/vs/platform/files/node/diskFileSystemProvider.ts` lists all the available file watchers:
- `UnixWatcherService` extends `Disposable`, and registers the new `Client`, so it is automatically disposed.
- `WindowsWatcherService` implements `IDisposable`, and in the `dispose()` function, kills the child process.
- `NsfwWatcherService` does the same thing as `UnixWatcherService`
- `NodeJSWatcherService` extends `Disposable`, and registers the `watchFile()` and `watchFolder()` calls. It calls `doWatchNonRecursive()` in `watcher.ts`. I can look into that code if necessary, but for now, I am assuming that function creates and disposes the resources correctly.
- `DiskFileSystemProvider` extends `Disposable`, but I just noticed that it might be leaked in a few cases. I'll create a new PR for this, because it affects a few other files.

bpasero replied:
Ah yeah good points 👍 . In fact, the disk file system provider is properly disposed from here:

https://github.com/microsoft/vscode/blob/8215a110a9469d99f42abe9dfcfa44e19282aa01/src/vs/workbench/electron-sandbox/shared.desktop.main.ts#L127-L127

I think the node.js watcher service - even if it were not disposed - is not an issue because that one does not use child processes.

bpasero commented:
Adding Alex for the changes in extension host.

rzhao271 commented:
@alexdima I've adjusted the `onWillShutdown()` and `onDidShutdown()` methods within `localProcessExtensionHost.ts`, because when I put a breakpoint within the `then` clause within `onDidShutdown()`, the process crashes. @deepak1556 is already looking into this, but considering how it is a long-running task anyway, I have moved it to `onWillShutdown()`.

I'm also wondering why there was a 10 second `setTimeout` used though? I'm not sure whether that can cause bugs considering how with process reuse, the workbench PID stays the same, but another extension host is launched.

alexdima reviewed (changes requested):
@rzhao271 

The extension host process can live after the renderer process terminates. The idea is that the renderer process sends a Terminate message to the extension host. The extension host will then call `deactivate` on extensions, collect their promises and wait up to 5 seconds for extensions to finish cleaning up and only then exit. The renderer had this very old fallback to kill the process after 10s. If I understand things correctly, the process would be long gone before that timeout even executed.

I think the current proposed change kills the extension host process immediately from the `onWillShutdown` event which is not correct. I suggest to simply remove the 10s extra careful timeout (just delete original lines 632-634) and leave the rest of the `terminate()` method as it was.

So even after adjusting this, I still have some concerns about the change with renderers reusing the same pid. In case the renderer would quit unexpectedly (without sending a Terminate message), the extension host would run every 1s a piece of code which checks if the renderer process is still alive. If the renderer process is gone, the extension host will unilaterally begin to exit (just as if the renderer process would have sent it a Terminate message).

The other more complex case is protecting against run-away `while(true)` extension host processes. We have a native node module called `native-watchdog` which is not very smart, but does the following: it spawns a thread in C++ and checks every 1s if the renderer process is alive or not. If the renderer process is not alive, the C++ code will brutally exit the extension host process in 7s or 10s, not sure the exact times. This serves as a watchdog for cases where extensions enter by accident `while(true)` loops. When the user closes the window, within 7-10s the extension host process will exit, even if the JS event loop is kept busy in an endless loop.

I am concerned that reusing the same process for the renderer will break these scenarios, so IMHO we need to find alternative implementations that don't rely on the renderer pid for this.

bpasero commented on src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts:
```diff
@@ -664,5 +653,6 @@ export class LocalProcessExtensionHost implements IExtensionHost {
 			this._extensionHostDebugService.terminateSession(this._environmentService.debugExtensionHost.debugId);
 			event.join(timeout(100 /* wait a bit for IPC to get delivered */), 'join.extensionDevelopment');
 		}
+		this.terminate();
```
Should this `terminate` call `join` the `onWillShutdown` to ensure the message is delivered before the app goes down? I think `terminate` itself runs a `promise.then` but does not signal that to the outside:

https://github.com/microsoft/vscode/blob/f250472adc449427b8b06ea77f23d780c982a40a/src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts#L624-L624

It also seems to me that `terminate` is called from more locations:
* https://github.com/microsoft/vscode/blob/f250472adc449427b8b06ea77f23d780c982a40a/src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts#L149-L151
* https://github.com/microsoft/vscode/blob/f250472adc449427b8b06ea77f23d780c982a40a/src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts#L142-L142

rzhao271 replied:
Locally, I created a new `terminate` function that returns a promise, and then joined it in `onWillShutdown`. I noticed that the reloading took noticeably longer, but I also managed to make it so that the extension host process isn't leaked. However, there's still some hangs that occur, so I need to look more into what's happening there.

rzhao271 commented:
@alexdima I talked with Deepak about the case where the render process quits unexpectedly, or where the watchdog detects that the extension host is unresponsive. We're thinking that the main process can spawn a watchdog process of sorts, and _that_ process can be used to monitor/kill the extension host in both cases. This way, we don't need the 1-second polling from the extension host, and we don't need the native watchdog to be loaded separately. Thoughts?

As for the case where the workbench is being reloaded, I've been trying a few implementations locally, but it seems that on Mac, unless I explicitly kill the extension host process using `this._extensionHostProcess.kill()`, the extension host always stays as a zombie process. I've tried adding a `this._extensionHostProcess.unref()` line right after the `fork()` call, so that the parent renderer process does not have to wait for the extension host to exit before proceeding with the reload, but that results in a zombie: on the next relaunch, the old extension host shows up in the process explorer as just `(Code - OSS Helper)`, and that process doesn't show up in the Activity Monitor. If I do wait for the extension host to exit using an `.on('exit', ...)` handler, and by joining this with the `onWillShutdown` event, the reload now takes much longer (up to 5 seconds longer), due to having to wait for the extension host to cleanly exit, but it doesn't show up as a zombie anymore.

bpasero commented on src/vs/workbench/services/search/electron-browser/searchService.ts:
```diff
@@ -84,6 +86,13 @@ export class DiskSearch implements ISearchResultProvider {
+		const disposeClient = async () => {
+			client.dispose();
+		};
+		this.lifecycleService.onWillShutdown((e) =>
```
If the disposal is not `async` you don`t have to do this joining on the event, just call the `dispose` from the event without `async` and without joining.

alexdima commented:
@deepak1556 @rzhao271 @bpasero I've pushed the changes we discussed about:
 - (1) the renderer process explicitly closes the renderer <-> extension host socket
 - (2) the extension host listens to the `close` event on the socket and behaves as if the renderer process asked it to terminate
 - I have rolled back all the other changes in this area

IMHO these changes (1) and (2) are good and I intend to also push them on `main`.

That being said, when using `"enable-render-process-reuse": true` in `argv.json` and doing a reload, I see that the socket change works fine; the extension host shutdown no longer depends on the renderer process id. In the extension host log, I see an entry e.g. `extension host terminating: renderer closed the socket`. So the extension host is able to realize that it needs to terminate, to `deactivate` extensions and exit.

But I'm still seeing what @rzhao271 saw. The extension host process is listed as a zombie process. e.g. in `htop`:

<img width="660" alt="Screen Shot 2021-05-05 at 16 46 51" src="https://user-images.githubusercontent.com/5047891/117161501-3fc3f680-adc2-11eb-9864-73d192e6fda0.png">

By looking at the log, I have validated that the extension host process has exited correctly. I'm not very knowledgeable about the inner workings of unix/linux processes, so I found this page helpful -- https://idea.popcount.org/2012-12-11-linux-process-states/

It appears that when the extension host process exits, it enters the zombie state until it is reaped by the parent process or until the parent process also exits. In the past, we never did `cp.kill()` on the extension host process because the extension host process can by design outlive the renderer process. But the extension host process never ended up in the zombie process, and I think that is because it would become "detached"/"orphaned". Now that the renderer process stays alive after a reload, the kernel expects that the renderer process will reap the extension host process. I think that means it needs to call `kill(extHostPid)`. And right now it doesn't do so and it is not possible for us to add a `.kill()` call before the reload, because that would defeat the design decision that the extension host process should outlive the renderer process.

So I think that we need to capture the `pid` of the extension host in the renderer process and communicate that `pid` to the reloaded renderer process and have the reloaded renderer process call `kill(pid)` like 10s after launching or something like that. Alternatively, the page at https://idea.popcount.org/2012-12-11-linux-process-states/ seems to describe a lot of ways in which zombie processes can be avoided by doing all kind of signal tricks in C++.

So I suggest three possibilities:
- figure out how to capture the `pid` of the extension host process in the old renderer, as it is shutting down and pass this `pid` to the new renderer (same process id) and ask the new renderer to call `kill(pid)` like after 10s or so.
- launch a C++ thread that is unrelated to nodejs / v8 from the renderer before it goes down that will call `kill(pid)` after 10s
- do some other C++ trick as described in the link above or maybe there are other ways to reap zombie processes.

bpasero commented:
@alexdima cool findings. We already have a method to kill a process by going through the main process:

https://github.com/microsoft/vscode/blob/16baed2dd411bb241fb259e67d37c84505cad5c7/src/vs/platform/native/electron-main/nativeHostMainService.ts#L484-L484

Can we not just add a `delay` option here and let the main process take care of killing the extension host after a while or does it really need to be the renderer process that kills the extension host?

deepak1556 commented:
Thanks for the great context @alexdima , to add some pointers from the internals. Whenever we call fork or spawn calls on the child_process module, they all hit the same codepath in libuv where it sets `SIGCHLD` handler https://github.com/libuv/libuv/blob/v1.x/src/unix/process.c#L431 to reap zombie processes when the child terminates https://github.com/libuv/libuv/blob/v1.x/src/unix/process.c#L75

The problem we are hitting is for whatever reason this `SIGCHLD` handler is not executed for the extension host that terminates after a timeout from the workbench reload, I am gonna check further on why that is the case.

> Can we not just add a delay option here and let the main process take care of killing the extension host after a while or does it really need to be the renderer process that kills the extension host?

@bpasero I don't think main process can help here, even if it sends a kill signal to the child extension host. The extension host will still expect the parent (which is workbench) to do a `waitpid` call.

> So I suggest three possibilities:

1) Sounds good idea where we still let the parent terminate the child but this would require us to have a handle to the old child process, I don't think we can do it after reload.

2) Will work but this requires us to deploy another native addon in the renderer, maybe this can be a last resort solution.

3) Would be to let libuv handle it as it normally should, I will look into this.

4) Having a detached process model sounds neat, let the extension host manage its lifetime would be useful when we move the host to be parented on other process when we sandbox the workbench. This is worth experimenting.

alexdima commented:
:+1: 

After thinking more about it, as we want to enable renderer sandboxing soon anyways, maybe I can look into moving the extension host spawning to the shared process.

bpasero commented:
💯 , happy to assist in that endeavour, maybe an architecture similar to how the PTY host is managing terminals in the shared process would work for extension host as well, given we will have 1 per window.

![image](https://user-images.githubusercontent.com/900690/117264510-9299be80-ae53-11eb-8f13-9cf571eb50a2.png)

rzhao271 commented:
I've rebased the PR.

TODO for myself:
- [ ] Confirm whether zombie processes are still created on the Mac.
- [ ] See whether `terminate` still has to be in the form of a promise, and whether we must still wait on it during `_onWillShutdown`.

rzhao271 commented:
I have confirmed that long-lasting zombie process are not created on macOS. Sometimes, a terminal process turns into a zombie, but it is reaped soon after.
However, on macOS, the main process seems to freeze very often, and its CPU usage goes up to around 100%. Moreover, the terminal often doesn't load with process reuse enabled, so I'm wondering whether there is something going on with the main-to-ptyHost communication path.

bpasero commented on src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts:
```diff
@@ -683,38 +683,30 @@ export class LocalProcessExtensionHost implements IExtensionHost {
-			// Give the extension host 10s, after which we will
-			// try to kill the process and release any resources
-			setTimeout(() => this._cleanResources(), 10 * 1000);
+	public terminate(): Promise<void> {
```
Are these changes specific to the renderer reuse flag or can they be ported to `main` independently if they are useful? Asking because now the extension host is not spawned from the workbench anymore so I wonder if they have purpose even now already?

rzhao271 replied:
Those changes were previously there to slow down `onWillShutdown` because otherwise,
1. Adding a breakpoint to that function and stepping through it would result in a crash, because the environment was being destroyed/shut down anyway.
2. If the resources weren't cleaned, a zombie process would occur on macOS.

Now with Alex's changes, point 2 is irrelevant. I believe point 1 still occurs, but considering how one has to add a breakpoint there to even repro the issue, it's unclear to me whether there is any problems with us leaving that code as-is. I have reverted that code, and now the PR only adds a runtime flag.

deepak1556 commented:
I would test process reuse with Electron 15 branch, I will push up the branch tomorrow. Electron 15 has far more fixes than 13 that cleans up the old process code path.

> on macOS, the main process seems to freeze very often, and its CPU usage goes up to around 100%

Can you take a profile and see where the slowdown is.

> Moreover, the terminal often doesn't load with process reuse enabled, so I'm wondering whether there is something going on with the main-to-ptyHost communication path.

Is it slow rendering or no rendering ? There is a know issue with the revive feature impacted by the shared process host https://github.com/microsoft/vscode/issues/133964, profile would be good understand what you are seeing.

rzhao271 commented:
I'm unable to repro the slowness issue or the terminal issue now. Considering how I was also seeing issues with the terminal, I think recent changes to the terminal fixed the issue.

deepak1556 reviewed (changes requested):
The flag should be opt-out and not for opt-in, as the default will be permanently set in https://github.com/microsoft/vscode/pull/137241

Also there will be spec failures related to `node-pty`, `sqlite3` and `spdlog`. Currently tracking those work in the above PR.

alexdima reviewed (approved):
This no longer contains any ext host changes, so :+1: from me.

rzhao271 commented:
Ben reminded me that the flag was not changeable for Electron >=14 anyway.
To me, it makes more sense for the setting to be opt-in for Electron 13 for easier merging + testing. After we're on Electron 15 stable, we can remove the lines having to do with `app.allowRendererProcessReuse` altogether.

Ref https://github.com/electron/electron/issues/18397
